rust/gelato/src/fwd_req_call.rs

```diff
@@ -0,0 +1,242 @@
+use crate::err::GelatoError;
+use crate::fwd_req_op::ForwardRequestOpArgs;
+use ethers::types::Signature;
+use serde::ser::SerializeStruct;
+use serde::{Deserialize, Serialize, Serializer};
+use std::sync::Arc;
+use tracing::info;
+use tracing::instrument;
+
+const GATEWAY_URL: &str = "https://gateway.api.gelato.digital";
+
+#[derive(Debug, Clone)]
+pub struct ForwardRequestCall {
+    pub http: Arc<reqwest::Client>,
+    pub args: ForwardRequestOpArgs,
+    pub sig: Signature,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ForwardRequestCallResult {
+    pub task_id: String,
+}
+
+impl ForwardRequestCall {
+    #[instrument]
+    pub async fn run(&self) -> Result<ForwardRequestCallResult, GelatoError> {
+        let url = format!(
+            "{}/metabox-relays/{}",
+            GATEWAY_URL,
+            u32::from(self.args.chain_id)
+        );
+        let http_args = HTTPArgs {
+            args: self.args.clone(),
+            sig: self.sig,
+        };
+        info!(?url, ?http_args);
+        let res = self.http.post(url).json(&http_args).send().await?;
+        let result = HTTPResult::from(res.json().await.unwrap());
+        Ok(ForwardRequestCallResult::from(result))
+    }
+}
+
+#[derive(Debug)]
+struct HTTPArgs {
+    args: ForwardRequestOpArgs,
+    sig: Signature,
+}
+
+#[derive(Debug, Clone, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+struct HTTPResult {
+    pub task_id: String,
+}
+
+impl Serialize for HTTPArgs {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("ForwardRequestHTTPArgs", 14)?;
+        state.serialize_field("typeId", "ForwardRequest")?;
+        state.serialize_field("chainId", &(u32::from(self.args.chain_id)))?;
+        state.serialize_field("target", &self.args.target)?;
+        state.serialize_field("data", &self.args.data)?;
+        state.serialize_field("feeToken", &self.args.fee_token)?;
+        // TODO(webbhorn): Get rid of the clone and cast.
+        state.serialize_field("paymentType", &(self.args.payment_type.clone() as u64))?;
+        state.serialize_field("maxFee", &self.args.max_fee.to_string())?;
+        state.serialize_field("gas", &self.args.gas.to_string())?;
+        state.serialize_field("sponsor", &self.args.sponsor)?;
+        // TODO(webbhorn): Just implement a `From<Chain> for H160` directly?
```

Comment:
not sure I follow?

Reply:
I think this was a stale TODO when I had a more convoluted type conversion originally.